data-plane/src/server/data_plane_server.rs

```diff
@@ -54,88 +56,353 @@ where
-                                            println!("Failed to send transaction context to log handler. err: {e}")
-                                        }
-                                    }
-                                },
-                                Err(e) => {
-                                    println!("Failed to build transaction context. err: {e:?}")
-                                }
-                            };
-
-                            let res: Result<Response<Body>> = Ok(response);
-                            res
                         }
-                    }),
-                )
-                .await;
-
-            if let Err(processing_err) = sent_response {
-                eprintln!("An error occurred while processing your request — {processing_err}");
+                    }
+                }
+                _ => {
+                    // We need to figure out a better auth mechanism for other protocols
+                    if feature_context.api_key_auth {
+                        eprintln!("API key auth needs to be turned off for non HTTPS/WS streams");
+                        shutdown_conn(&mut stream).await;
+                    }
+                    if let Err(err) = pipe_to_customer_process(&mut stream, &buffer, port).await {
+                        eprintln!("Failed piping non HTTP/WS stream to customer process — {err:?}");
+                        shutdown_conn(&mut stream).await;
+                    }
+                }
             }
         });
     }
 }
 
+async fn auth_request_non_http(
+    headers: &[httparse::Header<'_>],
+    e3_client: Arc<E3Client>,
+) -> Result<()> {
+    let api_key = headers
+        .iter()
+        .find(|header| header.name.to_ascii_lowercase() == "api-key")
+        .ok_or(MissingApiKey)?;
+    let header_value = HeaderValue::from_bytes(api_key.value)?;
+    if FeatureContext::get().api_key_auth
+        && auth_request(header_value, CageContext::get()?, e3_client)
+            .await
+            .is_some()
+    {
+        return Err(ApiKeyInvalid);
+    }
+    Ok(())
+}
+
+async fn shutdown_conn<L>(stream: &mut TlsStream<L>)
+where
+    TlsStream<L>: AsyncWriteExt + Unpin,
+{
+    if let Err(e) = stream.shutdown().await {
+        eprintln!("Failed to shutdown data plane connection — {e:?}");
+    }
+}
+
+async fn process_stream<L>(stream: &mut TlsStream<L>) -> Result<Vec<u8>>
+where
+    TlsStream<L>: AsyncReadExt + Unpin,
+{
+    let mut buffer = Vec::new();
+    loop {
+        let mut request_chunk = [0; 1024];
+        match stream.read(&mut request_chunk).await {
+            Ok(read_bytes) => {
+                if read_bytes == 0 {
+                    break;
+                }
+
+                buffer.extend_from_slice(&request_chunk[..read_bytes]);
+
+                if read_bytes < request_chunk.len() {
+                    break;
+                }
```

Comment:
I'm not sure this check will work, the only guarantee from tokio is `0 <= n <= buf.len()`. There's also a risk that an un-authed chunked payload could infinitely buffer data

Reply:
Ah ok what do you think is the best approach here for the check? I can add a request limit for unauthed reqs

Reply:
I think iteratively trying to parse the headers from the read bytes is the most reliable way to handle it